In [2]:
import pandas as pd
import numpy as np

# Display all columns
pd.set_option("display.max_columns", None)

# Display wider tables
pd.set_option("display.width", 120)

In [4]:
dataset_path = "../data/raw/online_retail.csv"

df = pd.read_csv(dataset_path)

print("Dataset Loaded Successfully!")

Dataset Loaded Successfully!


In [5]:
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2022-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2022-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2022-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2022-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2022-12-01 08:26:00,3.39,17850.0,United Kingdom


In [6]:
df.tail()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
541904,581587,22613,PACK OF 20 SPACEBOY NAPKINS,12,2023-12-09 12:50:00,0.85,12680.0,France
541905,581587,22899,CHILDREN'S APRON DOLLY GIRL,6,2023-12-09 12:50:00,2.10,12680.0,France
541906,581587,23254,CHILDRENS CUTLERY DOLLY GIRL,4,2023-12-09 12:50:00,4.15,12680.0,France
541907,581587,23255,CHILDRENS CUTLERY CIRCUS PARADE,4,2023-12-09 12:50:00,4.15,12680.0,France
541908,581587,22138,BAKING SET 9 PIECE RETROSPOT,3,2023-12-09 12:50:00,4.95,12680.0,France


In [7]:
rows, columns = df.shape

print(f"Number of Rows : {rows}")
print(f"Number of Columns : {columns}")

Number of Rows : 541909
Number of Columns : 8


In [8]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  str    
 1   StockCode    541909 non-null  str    
 2   Description  540455 non-null  str    
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  str    
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  str    
dtypes: float64(2), int64(1), str(5)
memory usage: 69.3 MB


In [9]:
df.describe()

,Quantity,UnitPrice,CustomerID
count,541909.000000,541909.000000,406829.000000
mean,9.552250,4.611114,15287.690570
std,218.081158,96.759853,1713.600303
min,-80995.000000,-11062.060000,12346.000000
25%,1.000000,1.250000,13953.000000
50%,3.000000,2.080000,15152.000000
75%,10.000000,4.130000,16791.000000
max,80995.000000,38970.000000,18287.000000


In [10]:
df.describe(include="all")

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
count,541909,541909,540455,541909.000000,541909,541909.000000,406829.000000,541909
unique,25900,4070,4223,NaN,23260,NaN,NaN,38
top,573585,85123A,WHITE HANGING HEART T-LIGHT HOLDER,NaN,2023-10-31 14:41:00,NaN,NaN,United Kingdom
freq,1114,2313,2369,NaN,1114,NaN,NaN,495478
mean,NaN,NaN,NaN,9.552250,NaN,4.611114,15287.690570,NaN
std,NaN,NaN,NaN,218.081158,NaN,96.759853,1713.600303,NaN
min,NaN,NaN,NaN,-80995.000000,NaN,-11062.060000,12346.000000,NaN
25%,NaN,NaN,NaN,1.000000,NaN,1.250000,13953.000000,NaN
50%,NaN,NaN,NaN,3.000000,NaN,2.080000,15152.000000,NaN
75%,NaN,NaN,NaN,10.000000,NaN,4.130000,16791.000000,NaN


In [11]:
print(df.columns.tolist())

['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'UnitPrice', 'CustomerID', 'Country']


In [12]:
df.dtypes

InvoiceNo          str
StockCode          str
Description        str
Quantity         int64
InvoiceDate        str
UnitPrice      float64
CustomerID     float64
Country            str
dtype: object

In [13]:
# Convert InvoiceDate to datetime
df["InvoiceDate"] = pd.to_datetime(df["InvoiceDate"])

print("InvoiceDate converted successfully.")

InvoiceDate converted successfully.


In [14]:
df.dtypes

InvoiceNo                 str
StockCode                 str
Description               str
Quantity                int64
InvoiceDate    datetime64[us]
UnitPrice             float64
CustomerID            float64
Country                   str
dtype: object

| Column      | Expected Data Type | Actual Data Type | Missing Values? | Initial Observation |
| ----------- | ------------------ | ---------------- | --------------- | ------------------- |
| InvoiceNo   | String             | ?                | ?               | ?                   |
| StockCode   | String             | ?                | ?               | ?                   |
| Description | String             | ?                | ?               | ?                   |
| Quantity    | Integer            | ?                | ?               | ?                   |
| InvoiceDate | Datetime           | ?                | ?               | ?                   |
| UnitPrice   | Float              | ?                | ?               | ?                   |
| CustomerID  | Integer            | ?                | ?               | ?                   |
| Country     | String             | ?                | ?               | ?                   |


In [16]:
# Count missing values
missing_values = df.isnull().sum()

# Calculate percentage of missing values
missing_percentage = (missing_values / len(df)) * 100

# Create a summary table
missing_summary = pd.DataFrame({
    "Missing Values": missing_values,
    "Percentage (%)": missing_percentage.round(2)
})

missing_summary

,Missing Values,Percentage (%)
InvoiceNo,0,0.00
StockCode,0,0.00
Description,1454,0.27
Quantity,0,0.00
InvoiceDate,0,0.00
UnitPrice,0,0.00
CustomerID,135080,24.93
Country,0,0.00


In [17]:
duplicate_count = df.duplicated().sum()

print(f"Total Duplicate Rows: {duplicate_count}")

Total Duplicate Rows: 5268


In [18]:
unique_values = pd.DataFrame({
    "Column": df.columns,
    "Unique Values": [df[col].nunique() for col in df.columns]
})

unique_values

,Column,Unique Values
0,InvoiceNo,25900
1,StockCode,4070
2,Description,4223
3,Quantity,722
4,InvoiceDate,23260
5,UnitPrice,1630
6,CustomerID,4372
7,Country,38


In [19]:
datatype_summary = pd.DataFrame({
    "Column": df.columns,
    "Data Type": df.dtypes.astype(str)
})

datatype_summary

,Column,Data Type
InvoiceNo,InvoiceNo,str
StockCode,StockCode,str
Description,Description,str
Quantity,Quantity,int64
InvoiceDate,InvoiceDate,datetime64[us]
UnitPrice,UnitPrice,float64
CustomerID,CustomerID,float64
Country,Country,str


In [20]:
invalid_quantity = df[df["Quantity"] <= 0]

print("Rows with Invalid Quantity:", len(invalid_quantity))

Rows with Invalid Quantity: 10624


In [21]:
invalid_price = df[df["UnitPrice"] <= 0]

print("Rows with Invalid Unit Price:", len(invalid_price))

Rows with Invalid Unit Price: 2517


In [22]:
cancelled_orders = df[df["InvoiceNo"].str.startswith("C")]

print("Cancelled Orders:", len(cancelled_orders))

Cancelled Orders: 9288


In [23]:
df.describe(include="all")

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
count,541909,541909,540455,541909.000000,541909,541909.000000,406829.000000,541909
unique,25900,4070,4223,NaN,NaN,NaN,NaN,38
top,573585,85123A,WHITE HANGING HEART T-LIGHT HOLDER,NaN,NaN,NaN,NaN,United Kingdom
freq,1114,2313,2369,NaN,NaN,NaN,NaN,495478
mean,NaN,NaN,NaN,9.552250,2023-07-04 13:34:57.156386,4.611114,15287.690570,NaN
min,NaN,NaN,NaN,-80995.000000,2022-12-01 08:26:00,-11062.060000,12346.000000,NaN
25%,NaN,NaN,NaN,1.000000,2023-03-28 11:34:00,1.250000,13953.000000,NaN
50%,NaN,NaN,NaN,3.000000,2023-07-19 17:17:00,2.080000,15152.000000,NaN
75%,NaN,NaN,NaN,10.000000,2023-10-19 11:27:00,4.130000,16791.000000,NaN
max,NaN,NaN,NaN,80995.000000,2023-12-09 12:50:00,38970.000000,18287.000000,NaN


| Issue               | Affected Rows | Business Risk            | Decision | Justification |
| ------------------- | ------------: | ------------------------ | -------- | ------------- |
| Missing CustomerID  |             ? | Cannot identify customer | Pending  | Investigate   |
| Missing Description |             ? | Product unknown          | Pending  | Investigate   |
| Duplicate Records   |             ? | Inflated sales           | Pending  | Investigate   |
| Cancelled Orders    |             ? | Not actual sales         | Pending  | Investigate   |
| Invalid Quantity    |             ? | Incorrect revenue        | Pending  | Investigate   |
| Invalid UnitPrice   |             ? | Incorrect revenue        | Pending  | Investigate   |


In [24]:
# Create a copy of the original dataset
clean_df = df.copy()

print("Original Shape:", clean_df.shape)

Original Shape: (541909, 8)


In [25]:
# Before cleaning
before_rows = clean_df.shape[0]

# Remove rows with missing CustomerID
clean_df = clean_df.dropna(subset=["CustomerID"])

# After cleaning
after_rows = clean_df.shape[0]

print(f"Rows Before: {before_rows}")
print(f"Rows After : {after_rows}")
print(f"Rows Removed: {before_rows - after_rows}")

Rows Before: 541909
Rows After : 406829
Rows Removed: 135080


In [26]:
before_rows = clean_df.shape[0]

clean_df = clean_df.dropna(subset=["Description"])

after_rows = clean_df.shape[0]

print(f"Rows Before: {before_rows}")
print(f"Rows After : {after_rows}")
print(f"Rows Removed: {before_rows - after_rows}")

Rows Before: 406829
Rows After : 406829
Rows Removed: 0


In [27]:
before_rows = clean_df.shape[0]

clean_df = clean_df.drop_duplicates()

after_rows = clean_df.shape[0]

print(f"Rows Before: {before_rows}")
print(f"Rows After : {after_rows}")
print(f"Rows Removed: {before_rows - after_rows}")

Rows Before: 406829
Rows After : 401604
Rows Removed: 5225


In [28]:
before_rows = clean_df.shape[0]

clean_df = clean_df[~clean_df["InvoiceNo"].str.startswith("C")]

after_rows = clean_df.shape[0]

print(f"Rows Before: {before_rows}")
print(f"Rows After : {after_rows}")
print(f"Rows Removed: {before_rows - after_rows}")

Rows Before: 401604
Rows After : 392732
Rows Removed: 8872


In [29]:
before_rows = clean_df.shape[0]

clean_df = clean_df[clean_df["Quantity"] > 0]

after_rows = clean_df.shape[0]

print(f"Rows Before: {before_rows}")
print(f"Rows After : {after_rows}")
print(f"Rows Removed: {before_rows - after_rows}")

Rows Before: 392732
Rows After : 392732
Rows Removed: 0


In [30]:
before_rows = clean_df.shape[0]

clean_df = clean_df[clean_df["UnitPrice"] > 0]

after_rows = clean_df.shape[0]

print(f"Rows Before: {before_rows}")
print(f"Rows After : {after_rows}")
print(f"Rows Removed: {before_rows - after_rows}")

Rows Before: 392732
Rows After : 392692
Rows Removed: 40


In [31]:
print("Final Dataset Shape:", clean_df.shape)

print("\nRemaining Missing Values:")
print(clean_df.isnull().sum())

print("\nRemaining Duplicate Rows:")
print(clean_df.duplicated().sum())

print("\nInvalid Quantity Rows:")
print((clean_df["Quantity"] <= 0).sum())

print("\nInvalid Unit Price Rows:")
print((clean_df["UnitPrice"] <= 0).sum())

print("\nCancelled Orders:")
print(clean_df["InvoiceNo"].str.startswith("C").sum())

Final Dataset Shape: (392692, 8)

Remaining Missing Values:
InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64

Remaining Duplicate Rows:
0

Invalid Quantity Rows:
0

Invalid Unit Price Rows:
0

Cancelled Orders:
0


In [32]:
print("=" * 60)
print("FINAL DATA VALIDATION REPORT")
print("=" * 60)

print("\nDataset Shape")
print(clean_df.shape)

print("\nMissing Values")
print(clean_df.isnull().sum())

print("\nDuplicate Rows")
print(clean_df.duplicated().sum())

print("\nNegative Quantity")
print((clean_df["Quantity"] <= 0).sum())

print("\nNegative Unit Price")
print((clean_df["UnitPrice"] <= 0).sum())

print("\nCancelled Orders")
print(clean_df["InvoiceNo"].str.startswith("C").sum())

FINAL DATA VALIDATION REPORT

Dataset Shape
(392692, 8)

Missing Values
InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64

Duplicate Rows
0

Negative Quantity
0

Negative Unit Price
0

Cancelled Orders
0


In [33]:
clean_df.info()

<class 'pandas.DataFrame'>
Index: 392692 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    392692 non-null  str           
 1   StockCode    392692 non-null  str           
 2   Description  392692 non-null  str           
 3   Quantity     392692 non-null  int64         
 4   InvoiceDate  392692 non-null  datetime64[us]
 5   UnitPrice    392692 non-null  float64       
 6   CustomerID   392692 non-null  float64       
 7   Country      392692 non-null  str           
dtypes: datetime64[us](1), float64(2), int64(1), str(4)
memory usage: 46.2 MB


In [34]:
clean_df["Revenue"] = clean_df["Quantity"] * clean_df["UnitPrice"]

In [35]:
clean_df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,Revenue
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2022-12-01 08:26:00,2.55,17850.0,United Kingdom,15.30
1,536365,71053,WHITE METAL LANTERN,6,2022-12-01 08:26:00,3.39,17850.0,United Kingdom,20.34
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2022-12-01 08:26:00,2.75,17850.0,United Kingdom,22.00
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2022-12-01 08:26:00,3.39,17850.0,United Kingdom,20.34
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2022-12-01 08:26:00,3.39,17850.0,United Kingdom,20.34


In [36]:
clean_df.to_csv(
    "../data/cleaned/online_retail_cleaned.csv",
    index=False
)

print("Clean dataset saved successfully.")

Clean dataset saved successfully.


In [37]:
saved_df = pd.read_csv(
    "../data/cleaned/online_retail_cleaned.csv"
)

print(saved_df.shape)

(392692, 9)


| Cleaning Step       | Rows Removed | Reason                          | Business Impact              |
| ------------------- | -----------: | ------------------------------- | ---------------------------- |
| Missing CustomerID  |         ____ | Customer identification missing | Required for RFM             |
| Missing Description |         ____ | Unknown product                 | Required for recommendation  |
| Duplicate Rows      |        5,268 | Prevent double counting         | Accurate sales               |
| Cancelled Orders    |         ____ | Not completed purchases         | Reflect real buying behavior |
| Invalid Quantity    |         ____ | Incorrect purchases/returns     | Correct revenue              |
| Invalid UnitPrice   |         ____ | Invalid revenue                 | Reliable monetary analysis   |


| Metric            | Before Cleaning |    After Cleaning |
| ----------------- | --------------: | ----------------: |
| Total Rows        |         541,909 |             _____ |
| Total Columns     |               8 | 9 (Revenue added) |
| Missing Values    |         Present |                 0 |
| Duplicate Rows    |           5,268 |                 0 |
| Invalid Quantity  |          10,624 |                 0 |
| Invalid UnitPrice |           2,517 |                 0 |
| Cancelled Orders  |           9,288 |                 0 |
